### ISSUE TITLE
error in expand segment(host)

### ISSUE THREAD

**Issue by jiangshuai1995** [greenplum-db/gpdb#5377]
### Greenplum version or build
PostgreSQL 9.1beta2 (Greenplum Database 6.0.0-alpha.0+dev.7352.g464db98 build dev-oss) on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11), 64-bit compiled on Aug  1 2018 03:24:35
### OS version and uname -a
Linux master 3.10.0-514.el7.x86_64 #1 SMP Tue Nov 22 16:42:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

when I want to  expand segment：

[ERROR]:-gpexpand failed. exiting...
Traceback (most recent call last):
  File "/usr/local/gpdb4/bin/gpexpand", line 2776, in main
    gpexpand_db_status = gpexpand.prepare_gpdb_state(logger, dburl, options)
  File "/usr/local/gpdb4/bin/gpexpand", line 1017, in prepare_gpdb_state
    gpexpand.get_gpdb_in_state(GPDB_STARTED, options)
  File "/usr/local/gpdb4/bin/gpexpand", line 1046, in get_gpdb_in_state
    GpStart.local('Start GPDB')
  File "/usr/local/gpdb4/lib/python/gppylib/commands/gp.py", line 757, in local
    cmd.run(validateAfter=True)
  File "/usr/local/gpdb4/lib/python/gppylib/commands/base.py", line 717, in run
    self.validate()
  File "/usr/local/gpdb4/lib/python/gppylib/commands/base.py", line 764, in validate
    raise ExecutionError("non-zero rc: %d" % self.results.rc, self)
ExecutionError: ExecutionError: 'non-zero rc: 2' occured.  Details: '$GPHOME/bin/gpstart -a -v'  cmd had rc=2 completed=True halted=False

but when i using  “$GPHOME/bin/gpstart -a -v” ，this command is OK.

kernel.shmmax = 500000000
kernel.shmmni = 4096
kernel.shmall = 4000000000
kernel.sem = 50100 128256000 50100 2560
kernel.sysrq = 1
kernel.core_uses_pid = 1
kernel.msgmnb = 65536
kernel.msgmax = 65536
kernel.msgmni = 2048
net.ipv4.tcp_syncookies = 1
net.ipv4.ip_forward = 0
net.ipv4.conf.default.accept_source_route = 0
net.ipv4.tcp_tw_recycle = 1
net.ipv4.tcp_max_syn_backlog = 4096
net.ipv4.conf.all.arp_filter = 1
net.ipv4.ip_local_port_range = 1025 65535
net.core.netdev_max_backlog = 10000
net.core.rmem_max = 2097152
net.core.wmem_max = 2097152
vm.overcommit_memory = 2

**Comment by skahler-pivotal** [comment]
Is it possible to provide the lines from the pg_log directory in the startup.log and gpdb-YYYY-MM-DD of the master during the time of the failed start?

**Comment by jiangshuai1995** [comment]
I’m sorry for so long to reply ,because the previous question cannot be reproduced. I can't provide the log.
But gpexpand failed too.

[gpadmin@master ~]$ gpexpand -i expandfile -D test -v
20180807:17:56:09:025677 gpexpand:master:gpadmin-[DEBUG]:-Obtaining master's port from master data directory
20180807:17:56:09:025677 gpexpand:master:gpadmin-[DEBUG]:-Read from postgresql.conf port=15432
20180807:17:56:09:025677 gpexpand:master:gpadmin-[DEBUG]:-Read from postgresql.conf max_connections=250
20180807:17:56:09:025677 gpexpand:master:gpadmin-[INFO]:-local Greenplum Version: 'postgres (Greenplum Database) 6.0.0-alpha.0+dev.7352.g464db98 build dev-oss'
20180807:17:56:09:025677 gpexpand:master:gpadmin-[DEBUG]:-Connecting to dbname='template1'
20180807:17:56:09:025677 gpexpand:master:gpadmin-[INFO]:-master Greenplum Version: 'PostgreSQL 9.1beta2 (Greenplum Database 6.0.0-alpha.0+dev.7352.g464db98 build dev-oss) on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11), 64-bit compiled on Aug  1 2018 03:24:35'
20180807:17:56:09:025677 gpexpand:master:gpadmin-[DEBUG]:-postmaster_pid_exists: True tmpfile_exists: True lockfile_exists: True netstat port: True  pid: 25383
20180807:17:56:09:025677 gpexpand:master:gpadmin-[INFO]:-Querying gpexpand schema for current expansion state
20180807:17:56:09:025677 gpexpand:master:gpadmin-[DEBUG]:-Connecting to dbname='test'
20180807:17:56:09:025677 gpexpand:master:gpadmin-[DEBUG]:-Connecting to dbname='test'
20180807:17:56:09:025677 gpexpand:master:gpadmin-[ERROR]:-gpexpand failed. exiting...
Traceback (most recent call last):
  File "/usr/local/gpdb4/bin/gpexpand", line 2776, in main
    gpexpand_db_status = gpexpand.prepare_gpdb_state(logger, dburl, options)
  File "/usr/local/gpdb4/bin/gpexpand", line 1023, in prepare_gpdb_state
    raise Exception('Error while trying to query the gpexpand schema: %s' % e)
Exception: Error while trying to query the gpexpand schema: FATAL:  database "test" does not exist

LOG said database does not exist, but it exist in fact.

[gpadmin@master gpAdminLogs]$ psql -p 15432 -d test
psql (9.1beta2)
Type "help" for help.

test=# select relname from pg_class limit 1;
  relname  
-----------
 pg_shadow
(1 row)

The gpdb-YYYY-MM-DD(log) as follows.

2018-08-07 18:05:54.834778 CST,"gpadmin","template1",p25867,th-346527680,"::1","1234",2018-08-07 18:05:54 CST,0,con15,,seg-1,,,,sx1,"WARNING","01000","time constraints added on superuser role",,,,,,,0,,"auth.c",3132,
2018-08-07 18:05:54.835832 CST,"gpadmin","template1",p25867,th-346527680,"::1","1234",2018-08-07 18:05:54 CST,0,con15,cmd1,seg-1,,,,sx1,"LOG","00000","statement: BEGIN",,,,,,"BEGIN",0,,"postgres.c",1600,
2018-08-07 18:05:54.835981 CST,"gpadmin","template1",p25867,th-346527680,"::1","1234",2018-08-07 18:05:54 CST,0,con15,cmd2,seg-1,,,,sx1,"LOG","00000","statement: SET CLIENT_MIN_MESSAGES='ERROR'",,,,,,"SET CLIENT_MIN_MESSAGES='ERROR'",0,,"postgres.c",1600,
2018-08-07 18:05:54.836048 CST,"gpadmin","template1",p25867,th-346527680,"::1","1234",2018-08-07 18:05:54 CST,0,con15,cmd3,seg-1,,,,sx1,"LOG","00000","statement: COMMIT",,,,,,"COMMIT",0,,"postgres.c",1600,
2018-08-07 18:05:54.836126 CST,"gpadmin","template1",p25867,th-346527680,"::1","1234",2018-08-07 18:05:54 CST,0,con15,cmd4,seg-1,,,,sx1,"LOG","00000","statement: BEGIN",,,,,,"BEGIN",0,,"postgres.c",1600,
2018-08-07 18:05:54.836190 CST,"gpadmin","template1",p25867,th-346527680,"::1","1234",2018-08-07 18:05:54 CST,0,con15,cmd5,seg-1,,,,sx1,"LOG","00000","statement: select version();",,,,,,"select version();",0,,"postgres.c",1600,
2018-08-07 18:05:54.836584 CST,"gpadmin","template1",p25867,th-346527680,"::1","1234",2018-08-07 18:05:54 CST,0,con15,cmd7,seg-1,,,,sx1,"LOG","00000","statement: SELECT typname, typlen FROM pg_type WHERE oid=25",,,,,,"SELECT typname, typlen FROM pg_type WHERE oid=25",0,,"postgres.c",1600,
2018-08-07 18:12:53.566739 CST,"gpadmin","template1",p26088,th-346527680,"::1","1518",2018-08-07 18:12:53 CST,0,con16,,seg-1,,,,sx1,"WARNING","01000","time constraints added on superuser role",,,,,,,0,,"auth.c",3132,
2018-08-07 18:12:53.567807 CST,"gpadmin","template1",p26088,th-346527680,"::1","1518",2018-08-07 18:12:53 CST,0,con16,cmd1,seg-1,,,,sx1,"LOG","00000","statement: BEGIN",,,,,,"BEGIN",0,,"postgres.c",1600,
2018-08-07 18:12:53.567964 CST,"gpadmin","template1",p26088,th-346527680,"::1","1518",2018-08-07 18:12:53 CST,0,con16,cmd2,seg-1,,,,sx1,"LOG","00000","statement: SET CLIENT_MIN_MESSAGES='ERROR'",,,,,,"SET CLIENT_MIN_MESSAGES='ERROR'",0,,"postgres.c",1600,
2018-08-07 18:12:53.568032 CST,"gpadmin","template1",p26088,th-346527680,"::1","1518",2018-08-07 18:12:53 CST,0,con16,cmd3,seg-1,,,,sx1,"LOG","00000","statement: COMMIT",,,,,,"COMMIT",0,,"postgres.c",1600,
2018-08-07 18:12:53.568112 CST,"gpadmin","template1",p26088,th-346527680,"::1","1518",2018-08-07 18:12:53 CST,0,con16,cmd4,seg-1,,,,sx1,"LOG","00000","statement: BEGIN",,,,,,"BEGIN",0,,"postgres.c",1600,
2018-08-07 18:12:53.568196 CST,"gpadmin","template1",p26088,th-346527680,"::1","1518",2018-08-07 18:12:53 CST,0,con16,cmd5,seg-1,,,,sx1,"LOG","00000","statement: select version();",,,,,,"select version();",0,,"postgres.c",1600,
2018-08-07 18:12:53.568627 CST,"gpadmin","template1",p26088,th-346527680,"::1","1518",2018-08-07 18:12:53 CST,0,con16,cmd7,seg-1,,,,sx1,"LOG","00000","statement: SELECT typname, typlen FROM pg_type WHERE oid=25",,,,,,"SELECT typname, typlen FROM pg_type WHERE oid=25",0,,"postgres.c",1600,


Thank you for your reply.

**Comment by skahler-pivotal** [comment]
Can you clear up something I am seeing in the output. It looks like you are using a gpexpand in this location /usr/local/gpdb4/bin/gpexpand and it looks like gpdb4 and the database itself says it is psql (9.1beta2)

Are you compiling a OSS version of GPDB and putting it in the gpdb4 directory? Is it possible you are using gpdb4 gpexpand instead of the one with the OSS build?

**Comment by jiangshuai1995** [comment]
gpdb4 doesn‘t  mean  greenplun version 4.X，because this clusters has 4 segments.

I think I found the cause of the problem. I installed two clusters on one host and used different ports. The env has led to a bit of confusion.

[gpadmin@master ~]$ ps -ef|grep postgres
gpadmin   3264     1  0 19:33 ?        00:00:00 /usr/local/gpdb4/bin/postgres -D /home/gpadmin/gpdata/gpmaster/gpseg-1 -p 15432 --gp_dbid=1 --gp_num_contents_in_cluster=4 --silent-mode=true --gp_contentid=-1 -E
gpadmin   3265  3264  0 19:33 ?        00:00:00 postgres: 15432, master logger process   
gpadmin   3268  3264  0 19:33 ?        00:00:00 postgres: 15432, checkpointer process   
gpadmin   3269  3264  0 19:33 ?        00:00:00 postgres: 15432, writer process   
gpadmin   3270  3264  0 19:33 ?        00:00:00 postgres: 15432, wal writer process   
gpadmin   3271  3264  0 19:33 ?        00:00:00 postgres: 15432, stats collector process   
gpadmin   3272  3264  0 19:33 ?        00:00:00 postgres: 15432, seqserver process   
gpadmin   3273  3264  0 19:33 ?        00:00:00 postgres: 15432, ftsprobe process   
gpadmin   3274  3264  0 19:33 ?        00:00:00 postgres: 15432, sweeper process   
gpadmin   3275  3264  0 19:33 ?        00:00:00 postgres: 15432, global deadlock detector process   
gpadmin   3337 25011  0 19:38 pts/0    00:00:00 grep --color=auto postgres
gpadmin  17691     1  0 8月03 ?       00:00:10 /usr/local/gpdb/bin/postgres -D /home/gpadmin/gpdata_c/gpsegment/gpseg_c0 -p 40000 --gp_dbid=2 --gp_num_contents_in_cluster=1 --silent-mode=true --gp_contentid=0 -c gp_role=utility
gpadmin  17692 17691  0 8月03 ?       00:00:01 postgres: 40000, logger process   
gpadmin  17695 17691  0 8月03 ?       00:00:04 postgres: 40000, checkpointer process   
gpadmin  17696 17691  0 8月03 ?       00:00:11 postgres: 40000, writer process   
gpadmin  17697 17691  0 8月03 ?       00:00:11 postgres: 40000, wal writer process   
gpadmin  17698 17691  0 8月03 ?       00:00:01 postgres: 40000, stats collector process   
gpadmin  17699 17691  0 8月03 ?       00:00:04 postgres: 40000, sweeper process   
gpadmin  17709     1  0 8月03 ?       00:00:03 /usr/local/gpdb/bin/postgres -D /home/gpadmin/gpdata_c/gpmaster/gpseg_c-1 -p 6432 --gp_dbid=1 --gp_num_contents_in_cluster=1 --silent-mode=true --gp_contentid=-1 -E
gpadmin  17710 17709  0 8月03 ?       00:02:35 postgres:  6432, master logger process   
gpadmin  17713 17709  0 8月03 ?       00:00:02 postgres:  6432, checkpointer process   
gpadmin  17714 17709  0 8月03 ?       00:00:08 postgres:  6432, writer process   
gpadmin  17715 17709  0 8月03 ?       00:00:07 postgres:  6432, wal writer process   
gpadmin  17716 17709  0 8月03 ?       00:01:18 postgres:  6432, stats collector process   
gpadmin  17717 17709  0 8月03 ?       00:00:00 postgres:  6432, seqserver process   
gpadmin  17718 17709  0 8月03 ?       00:00:00 postgres:  6432, ftsprobe process   
gpadmin  17719 17709  0 8月03 ?       00:00:02 postgres:  6432, sweeper process   
gpadmin  17720 17709  0 8月03 ?       00:00:03 postgres:  6432, global deadlock detector process

Although log said “Read from postgresql.conf port=15432” ，another cluster  affects the operation.

when I create database test in another cluster(port :6432) ,this error cannot be reproduced.

I think I still don't install two clusters...I might be wrong about the environment.

Thank you for your reply.
